x/sequencer/keeper/invariants.go

```diff
@@ -1,79 +1,169 @@
-		)
-
+// proposer and successor status' should be sensible
+func InvariantStatus(k Keeper) uinv.Func {
+	return uinv.AnyErrorIsBreaking(func(ctx sdk.Context) error {
+		var errs []error
 		rollapps := k.rollappKeeper.GetAllRollapps(ctx)
-		for _, rollapp := range rollapps {
-			proposer := k.GetProposer(ctx, rollapp.RollappId)
-			if !proposer.Bonded() {
-				broken = true
-				msg += "proposer is not bonded " + rollapp.RollappId + "\n"
+		for _, ra := range rollapps {
+			err := checkRollappStatus(ctx, k, ra.RollappId)
+			err = errorsmod.Wrapf(err, "rollapp: %s", ra.RollappId)
+			errs = append(errs, err)
+		}
+		for _, seq := range k.AllProposers(ctx) {
+			if !k.IsProposer(ctx, seq) {
+				errs = append(errs, fmt.Errorf("proposer in query is not proposer: %s", seq.Address))
 			}
-			successor := k.GetSuccessor(ctx, rollapp.RollappId)
-			if !successor.Bonded() {
-				broken = true
-				msg += "successor is not bonded " + rollapp.RollappId + "\n"
+		}
+		for _, seq := range k.AllSuccessors(ctx) {
+			if !k.IsSuccessor(ctx, seq) {
+				errs = append(errs, fmt.Errorf("successor in query is not successor: %s", seq.Address))
 			}
+		}
+		return errors.Join(errs...)
+	})
+}
 
+func checkRollappStatus(ctx sdk.Context, k Keeper, ra string) error {
+	proposer := k.GetProposer(ctx, ra)
+	if !proposer.Bonded() {
+		return errors.New("proposer not bonded")
+	}
+	successor := k.GetSuccessor(ctx, ra)
+	if !successor.Bonded() {
+		return errors.New("successor not bonded")
+	}
+	if !proposer.Sentinel() && proposer.Address == successor.Address {
+		return errors.New("proposer and successor are the same")
+	}
+	if !successor.Sentinel() && proposer.Sentinel() {
+		return errors.New("proposer is sentinel but successor is not")
+	}
+	all := k.RollappSequencers(ctx, ra)
+	bonded := k.RollappSequencersByStatus(ctx, ra, types.Bonded)
+	unbonded := k.RollappSequencersByStatus(ctx, ra, types.Unbonded)
+	if len(all) != len(bonded)+len(unbonded) {
+		return errors.New("sequencer by rollapp length is not equal to sum of bonded, and unbonded")
+	}
+	return nil
+}
+
+// module balance must correspond to sequencer stakes, and sequencer stakes should be sensible
+func InvariantTokens(k Keeper) uinv.Func {
+	return uinv.AnyErrorIsBreaking(func(ctx sdk.Context) error {
+		var errs []error
+
+		for _, seq := range k.AllSequencers(ctx) {
+			err := checkSeqTokens(ctx, seq, k)
+			err = errorsmod.Wrapf(err, "sequencer: %s", seq.Address)
+			errs = append(errs, err)
+		}
+
+		if err := errors.Join(errs...); err != nil {
+			return err
 		}
 
-		return sdk.FormatInvariant(
-			types.ModuleName, "sequencer-bonded",
-			msg,
-		), broken
+		total := sdk.NewCoin(k.bondDenom(ctx), sdk.ZeroInt())
+		for _, seq := range k.AllSequencers(ctx) {
+			total = total.Add(seq.TokensCoin())
+		}
+		// check module balance is equal
+		moduleAcc := k.accountK.GetModuleAccount(ctx, types.ModuleName)
+		balances := k.bankKeeper.GetAllBalances(ctx, moduleAcc.GetAddress())
+		if 1 < len(balances) {
```

Comment:
nit: maybe `len(balances) > 1` is better? 
`1 < len(balances)` broke my brain for a second haha

Reply:
haha the amount of times this style broke my brain also.. 🫠

Reply:
broskis
There is some research paper somewhere that says most people prefer the way I do it

Reply:
don't you learn in school the numbers start with zero at the left and positive infinity to the right 😄 
maybe hebrew is different
But you have no excuse @keruch

Reply:
I guess it's how you read it: left to right. Would you also say "two is less than how many apples I have"?

Reply:
Fair I guess peoples brains have a different preference, linguistic or spatial, when reading

Reply:
@danwt actually when i saw this `if`, i also imagined a straight line with numbers increasing to the right. but anyway as Alex pointed out, the other way around sounds more natural (at least to me) when you "read" the code.